datalad/metadata/extract.py

```diff
@@ -0,0 +1,665 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Run one or more metadata extractors on a dataset or file(s)"""
+
+__docformat__ = 'restructuredtext'
+
+from os import curdir
+import os.path as op
+import re
+import logging
+from six import (
+    binary_type,
+    string_types,
+    iteritems,
+)
+from collections import (
+    Mapping,
+)
+
+from datalad import cfg
+from datalad.interface.base import Interface
+from datalad.interface.base import build_doc
+from datalad.interface.results import (
+    get_status_dict,
+    success_status_map,
+)
+from datalad.interface.utils import eval_results
+from datalad.distribution.dataset import (
+    datasetmethod,
+    EnsureDataset,
+    require_dataset,
+)
+from datalad.support.annexrepo import AnnexRepo
+from datalad.support.param import Parameter
+from datalad.support.constraints import (
+    EnsureNone,
+    EnsureStr,
+    EnsureChoice,
+    EnsureBool,
+)
+from datalad.metadata.metadata import (
+    _get_metadatarelevant_paths,
+    get_metadata_type,
+)
+from datalad.metadata.definitions import version as vocabulary_version
+from datalad.utils import (
+    assure_list,
+    as_unicode,
+)
+from datalad.dochelpers import exc_str
+from datalad.log import log_progress
+
+# API commands needed
+from datalad.distribution.subdatasets import Subdatasets
+
+lgr = logging.getLogger('datalad.metadata.metadata')
+
+
+@build_doc
+class ExtractMetadata(Interface):
+    """Run one or more of DataLad's metadata extractors on a dataset or file.
+
+    The result(s) are structured like the metadata DataLad would extract
+    during metadata aggregation. There is one result per dataset/file.
+
+    Examples:
+
+      Extract metadata with two extractors from a dataset in the current directory
+      and also from all its files::
+
+        $ datalad extract-metadata -d . --source frictionless_datapackage --source datalad_core
+
+      Extract XMP metadata from a single PDF that is not part of any dataset::
+
+        $ datalad extract-metadata --source xmp Downloads/freshfromtheweb.pdf
+    """
+
+    _params_ = dict(
+        sources=Parameter(
+            args=("--source",),
```

Comment:
I don't see a clear suggestion. I read it as "ignore the confusion that comes from asking for `--type bids` and get a result that says `type: 'file'`. The config item `nativetype` is also weird at best (I think `datalad.metadata.nativetype` -> `datalad.metadata.extract` would better reflect what it actually triggers). I can agree with `source` not being semantically better.

I cannot really see the argument that there are config settings that do not match the name of a command arg. This is a command that you can explicitly tell what you want (it can also query the config, true, but it is not that a single config setting would determine what runs and what not -- the reality is (sadly) more complex).

That being said, the API breaks in any case because of `files` -> `path` -- which was simply an outlier that needs to be rectified. And again, I see little to no value in preserving consistency with a three-line first draft implementation of a command that is literally not used anywhere in the code base.

As we are having this argument frequently, I am starting to think that we should not allow any API addition that has proven itself in an extension for at least a year.

Reply:
I am ok to break API when it makes it things "better". But if "source" isn't better, why to break it?  if anyone processes those records internally, they know what to expect from "type".  Either "nativetype" is the best one (it is not the first one either iirc) - also not sure, but so far at least it had correspondence with "type" as specified in cmdline .  I must confess that I didn't use "types" option, so for me/scripts it doesn't matter. I am just saying that bringing more (not less) of the different terms meaning the same thing to solve some internal naming problem is suboptimal.  I was suggesting to avoid that by reverting back to "type(s)".